src/xbot/common/math/PID.java

```diff
@@ -151,27 +166,51 @@ public double calculate(double goal, double current,
     }
     
     /**
-     * Check for all conditions where the tolerance is above 0. Since tolerances default to
-     * -1, it will skip any unassigned ones.
+     * With the addition of a time threshold, "onTarget" needs to be updated every tick, since
+     * isOnTarget() should return true if all conditions are met EVEN IF it's the first time
+     * being called.
+     * 
+     * While that's unlikely due to our command infrastructure, it's still possible, and the PID
+     * class should be able to handle that scenario.
+     * 
+     * As a result, this checkIsOnTarget method was added, which does all the onTarget calculations
+     * during every calculate() call. the isOnTarget() method just returns a stored class-level variable.
      */
-    public boolean isOnTarget() {
-        
-        if (!checkErrorThreshold && !checkDerivativeThreshold) {
+    private void checkIsOnTarget() {
+        if (!checkErrorThreshold && !checkDerivativeThreshold && !checkOnTargetForDuration) {
             // No tolerances are enabled, but isOnTarget is being called anyway. We still need to return something.
             // In this case, we return FALSE, as it promotes robot action (the command using this will complete
             // its activity, even if it doesn't signal that it is done to allow other actions to proceed).
-            return false;
+            onTarget = false;
+            return;
         }
         
-        boolean isOnTarget = true;
+        onTarget = true;
         
         if (checkErrorThreshold) {
-            isOnTarget &= errorIsSmall;
+            onTarget &= errorIsSmall;
         }
         if (checkDerivativeThreshold) {
-            isOnTarget &= derivativeIsSmall;
+            onTarget &= derivativeIsSmall;
         }
         
-        return isOnTarget;
+        if (checkOnTargetForDuration) {
+            if(onTarget){
+                if(waitingToStabilize == false){
+                    waitingToStabilize = true;
+                    startTime = Timer.getFPGATimestamp();
+                    onTarget = false;
+                } else {
+                    onTarget =  Timer.getFPGATimestamp() - startTime > onTargetThreshold;
```

Comment:
Yeah, I checked.

From WPILib's Timer.class:

/**
   * Return the system clock time in seconds. Return the time from the FPGA hardware clock in
   * seconds since the FPGA started.
   *
   * @return Robot running time in seconds.
   */